### PR TITLE
[HigherOrderOp] fallthrough some keys by default.

### LINKED ISSUE #109253: [HigherOrderOps] Default DispatchKey behavior
## Issue description

The HOO dispatching mechanism is separate from the C++ PyTorch dispatcher. Instead, it "simulates" dispatch in Python. We chose this design because it was easy to implement, we didn't want to teach the C++ dispatcher about the crazy things we are doing, and it worked.

In the PyTorch dispatcher, we are able to register a fallback for each DispatchKey. The fallback is a "default implementation" for the DispatchKey for each operator.

We should also simulate this in Python. Currently, each HOO manually registers a fallthrough on each DispatchKey; we should make it so that this happens automatically:

https://github.com/pytorch/pytorch/blob/faa5985dfefacc31485eb6fdec52ed62196d7a17/torch/_higher_order_ops/cond.py#L478-L484

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110478

Fixes #109253

Test Plan:
Added a new test that shows default fallthrough keys can be overrided.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zou3519

### CODE REVIEW

File: torch/_ops.py [hunk 1]
```diff
@@ -234,6 +234,15 @@ def resolve_key(op: OperatorBase, k: DispatchKey):  # type: ignore[valid-type]
 _global_higher_order_ops = {}
 _higher_order_ops = {}
 
+_HIGHER_ORDER_OP_DEFAULT_FALLTHROUGH_DISPATCH_KEYS = [
+    DispatchKey.PythonDispatcher,  # type: ignore[attr-defined]
+    DispatchKey.PythonTLSSnapshot,  # type: ignore[attr-defined]
+    DispatchKey.ADInplaceOrView,
+    DispatchKey.BackendSelect,
+    DispatchKey.AutocastCPU,  # type: ignore[attr-defined]
+    DispatchKey.AutocastCUDA,  # type: ignore[attr-defined]
```
Comment (zou3519) [inline torch/_ops.py:243]: Is one still able to py_impl for one of these if they wanted to? (could we add a test for that?)

Comment (ydwu4) [reply torch/_ops.py:243]: yes, I make HigherOrderOperator override the py_impl method of OperatorBase and will add it to the non_fallthrough_keys if it's in the fallthrough keys list. 

I can add a test.

**Review by ezyang — APPROVED** [review]
the implementation is dorky but whatever, it's easy enough to change later

**Pull request closed without merging**